File: tests/journal.lib/error_example.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the journal
#include <p2/journal.h>
// support
#include <cassert>


// type alias
using myerror_t = pyre::journal::error_t;


// exercise the channel manipulators
int main() {
    // make an error channel
    myerror_t channel("tests.journal.error");

    // deactivate the channel
    channel.deactivate();

    // carefully
    try {
        // inject something into the channel
        channel
            << pyre::journal::at(__HERE__)
            << pyre::journal::set("time", "now")
            << "error channel:" << pyre::journal::newline
            << "    hello world!" << pyre::journal::endl;
        // errors are fatal by default, so we shouldn't be able to get here
        throw std::logic_error("unreachable");
    // if all goes well
    }  catch (const myerror_t::exception_type & error) {
        // make sure the reason was recorded correctly
        assert (error.what() == channel.name() + myerror_t::string_type(": application error"));
    }

    // all done
    return 0;
}


// end of file

File: tests/journal.pkg/error_example.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


def test():
    """
    Exercise the error channel with a realistic example
    """
    # access the journal
    import j2

    # make an error channel
    channel = j2.error(name="tests.journal.error")
    # send the output to trash
    channel.device = j2.trash()

    # add some metadats
    channel.meta["time"] = "now"
    # inject
    channel.line("error channel:")
    channel.log("    hello world!")

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file
